Material component documentation (#1639)
o3de/o3de.org: content/docs/user-guide/components/reference/atom/material.md

---
title: Material Component
linktitle: Material
description: 'Open 3D Engine (O3DE) Material Component reference.'
toc: true
---

> The **Material** component is used to inspect, customize, and otherwise override materials and their properties on entities with compatible renderable components. [Mesh Component](/docs/user-guide/components/reference/atom/mesh/) and [Actor Component](/docs/user-guide/components/reference/atom/actor/) are currently the only components affected by the Material Component. However, any component can support the Material Component by implementing the required buses and services. The material assets used by the Material Component are most commonly created using the Material Editor. They are also produced by the asset processor and model asset builder, which exports the materials defined within a source model, like an FBX file. 

Reviewer: The Actor component link is broken.

toc: true
---

> Use the **Material** component to inspect, customize, and override materials and properties used by other compatible components. Currently, only the [Mesh component](/docs/user-guide/components/reference/atom/mesh/) and the [Actor component](/docs/user-guide/components/reference/animation/actor/) support the Material component. However, any component can support the Material component by implementing the required buses and services. The Material component uses material assets, which you can create and configure using the Material Editor. Another way that material assets are produced is by the Asset Processor and the Model Asset Builder. This exports the materials defined within a source model, like an FBX file.

Reviewer 2: ```suggestion
Use the **Material** component to inspect, customize, and override materials and properties on entities with compatible render components. Currently, only the [Mesh component](/docs/user-guide/components/reference/atom/mesh/) and the [Actor component](/docs/user-guide/components/reference/animation/actor/) support the Material component. However, any component can support the Material component by implementing the required buses and services. The Material component uses material assets, which you can create and configure using the Material Editor. Another way that material assets are produced is by the Asset Processor and the Model Asset Builder. This exports the materials defined within a source model, like an FBX file.
```

> The **Material** component is used to inspect, customize, and otherwise override materials and their properties on entities with compatible renderable components. [Mesh Component](/docs/user-guide/components/reference/atom/mesh/) and [Actor Component](/docs/user-guide/components/reference/animation/actor/) are currently the only components affected by the Material Component. However, any component can support the Material Component by implementing the required buses and services. The material assets used by the Material Component are most commonly created using the Material Editor. They are also produced by the asset processor and model asset builder, which exports the materials defined within a source model, like an FBX file. 

Reviewer 2: Ah yes that's key, my latest suggestion will address that.

| Complex | ![material-component-extended-slot-properties](/images/user-guide/components/reference/atom/material/material-extended-slot-properties-ui.png) |

## Material Component Features

| Description | Screenshot |
|-|-|
| The Material Component context menu contains many options for clearing all materials, model materials, and LOD materials. The menu has actions for eliminating unused material assignments and properties that might remain from a previously assigned model or if the Material Component is copied and pasted to a new entity. Most of these options correspond to functions on the MaterialComponentRequestBus detailed below. | ![material-componen-menu](/images/user-guide/components/reference/atom/material/material-menu-ui.png) |
| There is also a context menu available for each material slot. This menu gives options for opening the dedicated Material Editor, opening Material Instance Inspector windows for customizing material properties on selected entities, and more.  | ![material-component-slot-menu](/images/user-guide/components/reference/atom/material/material-slot-menu-ui.png) |

> | The context menus and large button at the top of the Material Component give access to the 'Generate/Manage Source Materials' dialog. The dialog displays a list of the unique materials associated with each material slot by default. After configuring target filenames, the dialog will be able to generate and automatically assign source material files to every matching material slot. The source materials can then be opened and edited with the Material Editor as well as shared with other entities. This window predates the existence of the Material Instance Inspector, the ability to customize materials without assigning a material asset, and the more robust model exporter that exists today. Using the dialog is not required but can simplify automatically assigning materials shared by multiple entities. It can also be useful for building a material library based on materials embedded in a FBX or other source. If source materials are not required for your workflow then materials can be customized and saved directly on entities using the Material Instance Inspector. | ![material-componen-generate](/images/user-guide/components/reference/atom/material/material-generate-ui.png) |

Author: It initializes with default filenames based on the material slot name. Users can optionally select their own file names. We may replace this workflow with one that just selects a folder for all of the materials.

Which workflow? If the materials were previously exported from this dialog and customized in the material editor then they can be applied to any other entity. If multiple entities are selected when this dialog is opened from the component context menu, and they all share the same set of material slots, the generated materials will be assigned on every entity.

> | The context menus and large button at the top of the Material Component give access to the 'Generate/Manage Source Materials' dialog. The dialog displays a list of the unique materials associated with each material slot by default. After configuring target filenames, the dialog will be able to generate and automatically assign source material files to every matching material slot. The source materials can then be opened and edited with the Material Editor as well as shared with other entities. This window predates the existence of the Material Instance Inspector, the ability to customize materials without assigning a material asset, and the more robust model exporter that exists today. Using the dialog is not required but can simplify automatically assigning materials shared by multiple entities. It can also be useful for building a material library based on materials embedded in a FBX or other source. If source materials are not required for your workflow then materials can be customized and saved directly on entities using the Material Instance Inspector. | ![material-componen-generate](/images/user-guide/components/reference/atom/material/material-generate-ui.png) |

Reviewer 2: That answers it perfectly, thanks!

| The Material Instance Inspector provides all of the same features as the inspector found in the Material Editor. Instead of creating a child material, you have the option of customizing material properties that override values from the asset assigned to a material slot. The changes will be applied to the entity in real time, visible in the viewport and previews. This window will remain opened and locked to the entity and material slot selection until manually closed. The UI will be disabled if the entity or the selected material slot become invalid or unavailable. | ![material-componen-inspector](/images/user-guide/components/reference/atom/material/material-inspector-ui.png) |

| Property | Description | Values | Default |
|-|-|-|-|

> | **Generate/Manage Source Materials...** | This button opens the "Generate/Manage Source Materials" dialog. | | |

Reviewer 2: ```suggestion
| **Generate/Manage Source Materials...** | This button opens the **Generate/Manage Source Materials** dialog. | | |
```

> | **Default Material** | Materials assigned to the default material slot will be applied to the entire model wherever no higher priority model or LOD material is assigned. | Material Asset | None |

Reviewer 2: ```suggestion
| **Default Material** | The default material applies to the entire model, except for parts where a higher priority model or LOD material is assigned. | Material Asset | None |
```

> | **Model Materials** | Materials assigned to these slots will be applied to every part of the model with the same material slot name unless a higher priority LOD material is assigned. | Container of material assets | Empty |

Reviewer 2: ```suggestion
| **Model Materials** | Materials that are assigned to these slots are applied to every part of the model with the same material slot name, unless a higher priority LOD material is assigned. | Container of material assets | Empty |
```

| **Enable LOD Materials** | When this flag is enabled, materials can be customized per LOD. | Boolean | `Disabled` |

> | **LOD Materials** | Materials assigned to LOD material slots will override model and default material assignments for the given LOD. If you are changing the default material or model materials and not seeing those changes applied then verify that there is no LOD material taking higher priority or disable them completely. | Container of material assets organized by LOD | Empty |

Reviewer 2: ```suggestion
| **LOD Materials** | Materials assigned to LOD material slots override **Default Material** and **Model Material** assignments for the given LOD. If you change the default material or model materials and don't see those changes applied, then verify that there's no LOD material taking higher priority or disable them completely. | Container of material assets organized by LOD | Empty |
```

## MaterialComponentRequestBus

> This request bus supplies the interface for interacting with the Material Component. Many of the functions relate to inspecting or customizing material asset and property overrides. All of the functions are available in the editor, gameplay, and simulation. Many of them may be more useful for extending Material Component related tools and scripts. 

Reviewer 2: ```suggestion
This request bus supplies the interface for interacting with the Material component. It contains many of the functions that relate to inspecting or customizing material asset and property overrides. All of the functions are available in the editor, gameplay, and simulation. Many of them may be more useful for extending Material component related tools and scripts. 
```

| Method Name | Description | Parameter | Return | Scriptable |
|-|-|-|-|-|

> | 'GetOriginalMaterialAssignments' | Returns a map of available material slots and their default assignments | None | Unmodified Material Slot Layout: MaterialAssignmentMap | Yes |

Reviewer 2: ```suggestion
| 'GetOriginalMaterialAssignments' | Returns a map of available material slots and their default assignments. | None | Unmodified Material Slot Layout: MaterialAssignmentMap | Yes |
```

Missing a period here and other method descriptions.
